R/florabank_queries.R

```diff
@@ -0,0 +1,424 @@
+#' Omschrijving en Rekenwaarde. The first four variables identify the taxon,
+#' the latter four variables relate to the taxon traits.
+#'
+#' @importFrom dplyr
+#' tbl
+#' collect
+#' distinct
+#' pull
+#' %>%
+#' inner_join
+#' left_join
+#' filter
+#' select
+#' rename
+#' @importFrom rlang .data
+#' @importFrom assertthat assert_that
+#'
+#' @export
+#'
+#' @examples
+#' \dontrun{
+#' # connect to florabank
+#' db_connectie <- connect_inbo_dbase("D0021_00_userFlora")
+#'
+#' # get all Ellenberg values via partial matching
+#' fb_ellenberg <- florabank_traits(db_connectie, "llenberg")
+#'
+#' # get all red lists via partial matching
+#' fb_rodelijsten <- florabank_traits(db_connectie, "rode")
+#'
+#' # get only the red list for vascular plant species
+#' fb_rodelijstvaatplanten <- florabank_traits(db_connectie, "Rode lijst Vaatplanten")
+#'
+#' #if the trait_name argument is missing, a list of possible names is printed
+#' florabank_traits(db_connectie)
+#'
+#' #disconnect from florabank
+#' dbDisconnect(db_connectie)
+#' }
+
+florabank_traits <- function(connection, trait_name) {
+
+  assert_that(class(connection)[1] == "Microsoft SQL Server",
+              msg = "Not a connection object to database.")
+  assert_that(connection@info$dbname == "D0021_00_userFlora")
+
+  if (missing(trait_name)) {
+    traitnames <- tbl(connection, "tblTaxonKenmerk") %>%
+      distinct(.data$Naam) %>%
+      collect() %>%
+      pull(.data$Naam)
+    message(paste(traitnames, collapse = ", "))
+    stop("Please provide (part of) a trait name from the above list.")
+  }
+
+  trait_name <- tolower(trait_name)
+
+  fb_taxon <- tbl(connection, "tblTaxon")
+  fb_taxon_kenmerk <- tbl(connection, "tblTaxonKenmerk")
+  fb_taxon_kenmerk_waarde <- tbl(connection, "tblTaxonKenmerkWaarde")
+  rel_taxon_taxon_kenmerk_waarde <- tbl(connection, "relTaxonTaxonKenmerkWaarde")
+
+  query_result <- rel_taxon_taxon_kenmerk_waarde %>%
+    inner_join(fb_taxon_kenmerk %>%
+                 filter(tolower(.data$Naam) %LIKE%
+                          paste0("%", trait_name, "%")) %>%
+                 select(.data$ID, .data$Naam),
+               by = c("TaxonKenmerkID" = "ID")) %>%
+    select(-.data$Omschrijving) %>%
+    left_join(fb_taxon_kenmerk_waarde %>%
+                distinct(.data$ID, .data$Code, .data$TaxonKenmerkID,
+                         .data$Omschrijving, .data$Rekenwaarde),
+              by = c("TaxonKenmerkID" = "TaxonKenmerkID",
+                     "TaxonKenmerkWaardeID" = "ID")) %>%
+    left_join(fb_taxon %>%
+                rename(NaamAfkorting = .data$Code),
+              by = c("TaxonID" = "ID")) %>%
+    distinct(.data$TaxonID,
+             TaxonAfkorting = .data$NaamAfkorting,
+             TaxonWetenschappelijk = .data$NaamWetenschappelijk,
+             TaxonNederlands = .data$NaamNederlands,
+             Kenmerk = .data$Naam,
+             .data$Code,
+             .data$Omschrijving,
+             .data$Rekenwaarde
+    ) %>%
+    collect()
```

Comment:
Maybe don't add `collect()` and leave that for the user?

Reply:
Resolved in c9f25d1